Optimize String#casecmp

dan42 opened the pull request:
I was inspired by PR #3369 to play around with casecmp optimization.

I split these optimizations to str_casecmp in two independent commits, for ease of review.

1. In a tight loop, each call to rb_enc_ascget() was calling rb_enc_asciicompat() which always returns te same value. So I unrolled that. Yields up to 4x speedup.

2. For ascii-compatible strings, check for ascii case-equality 8 chars at a time by using vectors and bit arithmetic. Yields up to 6x speedup.

Combined, these two give a **12x** speedup for the casecmp-str-100 benchmark!

Also changed `benchmark/string_casecmp.yml` to properly test the various code branches.


|                   |compare-ruby|built1-ruby|built2-ruby|
|:------------------|-----------:|----------:|----------:|
|casecmp-ascii-1    |      4.515M|     4.497M|     6.201M|
|                   |       1.00x|          -|      1.38x|
|casecmp-ascii-10   |    995.791k|   990.196k|     3.494M|
|                   |       1.01x|          -|      3.53x|
|casecmp-ascii-100  |    114.624k|   112.729k|   650.711k|
|                   |       1.02x|          -|      5.77x|
|casecmp-ascii-x    |      4.518M|     4.521M|     6.154M|
|                   |           -|      1.00x|      1.36x|
|casecmp-latin-1    |      4.164M|     4.147M|     5.623M|
|                   |       1.00x|          -|      1.36x|
|casecmp-latin-10   |    935.471k|   943.667k|     2.721M|
|                   |           -|      1.01x|      2.91x|
|casecmp-latin-100  |    107.396k|   108.505k|   431.820k|
|                   |           -|      1.01x|      4.02x|
|casecmp-latin-x    |      4.091M|     4.071M|     5.518M|
|                   |       1.01x|          -|      1.36x|
|casecmp-str-1      |      1.557M|     3.730M|     4.548M|
|                   |           -|      2.40x|      2.92x|
|casecmp-str-10     |    194.046k|   744.211k|     1.861M|
|                   |           -|      3.84x|      9.59x|
|casecmp-str-100    |     19.970k|    83.117k|   254.051k|
|                   |           -|      4.16x|     12.72x|
|casecmp-str-x      |      1.468M|     3.473M|     4.578M|
|                   |           -|      2.37x|      3.12x|
|casecmp-wide-1     |      1.103M|     1.287M|     1.288M|
|                   |           -|      1.17x|      1.17x|
|casecmp-wide-10    |    129.433k|   155.519k|   154.168k|
|                   |           -|      1.20x|      1.19x|
|casecmp-wide-100   |     13.149k|    15.884k|    15.893k|
|                   |           -|      1.21x|      1.21x|
|casecmp-wide-x     |      1.055M|     1.267M|     1.225M|
|                   |           -|      1.20x|      1.16x|

nobu commented on string.c:
```diff
@@ -3463,6 +3463,98 @@ rb_str_cmp_m(VALUE str1, VALUE str2)
+#endif
+
+static long
+ascii_eqlcase_len(char *p1, char *p2, long len, long len2)
+{
+    register uintptr_t v1, v2;
+    uintptr_t bits;
+    if (len2 < len) len = len2;
+    char *p1initial = p1, *p1end;
```
```suggestion
ascii_eqlcase_len(const char *p1, const char *p2, long len, long len2)
{
    register uintptr_t v1, v2;
    uintptr_t bits;
    if (len2 < len) len = len2;
    const char *p1initial = p1, *p1end;
```

dan42 commented:
* rebased to master
* incorporared the previous fixes into the proper commits
* fixed the UTF-16LE ordering while pending a decision on https://bugs.ruby-lang.org/issues/17594
* added an additional optimization for ascii-safe encodings; this brings the best-case speedup to **27x!**

dan42 commented on encoding.c:
```diff
@@ -386,6 +393,25 @@ enc_register_at(struct enc_table *enc_table, int index, const char *name, rb_enc
     }
     encoding->name = name;
     encoding->ruby_encoding_index = index;
+    if (encoding->max_enc_len == 1 ||
+        STRCASECMP(name, "UTF-8") == 0 ||
+        STRCASECMP(name, "UTF8-MAC") == 0 ||
+        STRCASECMP(name, "UTF8-DoCoMo") == 0 ||
+        STRCASECMP(name, "UTF8-KDDI") == 0 ||
+        STRCASECMP(name, "UTF8-SoftBank") == 0 ||
+        STRCASECMP(name, "CP51932") == 0 ||
+        STRCASECMP(name, "Emacs-Mule") == 0 ||
+        STRCASECMP(name, "EUC-JIS-2004") == 0 ||
+        STRCASECMP(name, "EUC-JP") == 0 ||
+        STRCASECMP(name, "eucJP-ms") == 0 ||
+        STRCASECMP(name, "EUC-KR") == 0 ||
+        STRCASECMP(name, "EUC-TW") == 0 ||
+        STRCASECMP(name, "GB12345") == 0 ||
+        STRCASECMP(name, "GB2312") == 0 ||
+        STRCASECMP(name, "stateless-ISO-2022-JP") == 0 ||
+        STRCASECMP(name, "stateless-ISO-2022-JP-KDDI") == 0) {
+        encoding->flags |= ENC_ASCIISAFE_FLAG;
```
I don't think this is the proper place to hardcode the list of ascii-safe encoding, but I really didn't know where else to put it.

Also, while ENC_DUMMY_FLAG is stored in `ruby_encoding_index`, I store the ENC_ASCIISAFE_FLAG bit in `flags` instead. It just seemed to make more sense?

dan42 commented:
@nobu may I ask if there's a problem that is blocking this PR?

mame commented on encoding.c:
```diff
@@ -89,6 +89,7 @@ static rb_encoding *global_enc_ascii,
 
 #define ENC_DUMMY_FLAG (1<<24)
+#define ENC_ASCIISAFE_FLAG (1<<25)
 #define ENC_INDEX_MASK (~(~0U<<24))
```
Should `ENC_INDEX_MASK` be updated to `(~(~0U<<25))`?

dan42 replied:
No, that's the mask for bits 0..23 which are used to store the index in `ruby_encoding_index`

nurse commented:
> Improve casecmp benchmark to cover more encodings

I don't think ISO-8859-1 and UTF-16's cases are worth to add.

> Optimize String#casecmp by extracting condition from loop

Adding `else if (rb_enc_asciicompat(enc)) {`  case sounds reasonable.

Optimizing `else {` case (non ASCII compatible encoding) is not worth optimizing.
memcmp2 is bad name.
rb_enc_mbcget is not worth introducing for this.

> Add flag for ascii-safe encodings

The idea "ascii-safe encoding" is not so bad.
But I think just supporting US-ASCII, ASCII-8BIT, and UTF-8 is good enough for real world application.

> Optimize String#casecmp with vector bit arithmetic

This is too complicated to merge without evaluating real world impact. You need to show how large this optimization improves real world application's performance.

dan42 commented:
> I don't think ISO-8859-1 and UTF-16's cases are worth to add.

The idea was to test the different branches of the code. Originally ISO-8859-1 and ASCII had different performance characteristics but they are now the same, so ISO-8859-1 can be removed. But I still think it makes sense to test the 3 branches of `str_casecmp`.

> Optimizing `else {` case (non ASCII compatible encoding) is not worth optimizing.

I understand the `else` branch is almost never used... it just feels silly to me that `rb_enc_ascget` will check the value of `rb_enc_asciicompat` for every char when we know it will always be false. And call `ONIGENC_PRECISE_MBC_ENC_LEN` twice for each char.

> memcmp2 is bad name.

The name means "memcmp" with "2" lengths instead of 1. Do you have a better name in mind? Or I could just leave that code inline, duplicated in the two places where `memcmp2` is used.

> rb_enc_mbcget is not worth introducing for this.

It's actually almost the same as `rb_enc_codepoint_len` except it returns -1 instead of raising an error. I was surprised there wasn't already a function for that.

> The idea "ascii-safe encoding" is not so bad.
> But I think just supporting US-ASCII, ASCII-8BIT, and UTF-8 is good enough for real world application.

I don't understand the purpose of limiting this.
I would not go out of my way to optimize EUC-JP, but if it can benefit from the same optimizations used for UTF-8, then why not?
Maybe you mean that instead of a flag bit we can just use `enc->max_enc_len == 1 || enc == global_enc_utf_8` ?

> > Optimize String#casecmp with vector bit arithmetic
> 
> This is too complicated to merge without evaluating real world impact. You need to show how large this optimization improves real world application's performance.

Good point, maybe `casecmp` performance is just not so important overall.
Do you have any suggestions on how to test that?
